org.lflang/src/org/lflang/generator/ts/TsGenerator.kt

```diff
@@ -0,0 +1,436 @@
+ *  @author{Edward A. Lee <eal@berkeley.edu>}
+ *  @author{Marten Lohstroh <marten@berkeley.edu>}
+ *  @author {Christian Menard <christian.menard@tu-dresden.de>}
+ *  @author {Hokeun Kim <hokeunkim@berkeley.edu>}
+ */
+class TsGenerator(
+    private val tsFileConfig: TsFileConfig,
+    errorReporter: ErrorReporter,
+    private val scopeProvider: LFGlobalScopeProvider
+) : GeneratorBase(tsFileConfig, errorReporter) {
+
+    companion object {
+        /** Path to the Cpp lib directory (relative to class path)  */
+        const val LIB_PATH = "/lib/TS"
+
+        /**
+         * Names of the configuration files to check for and copy to the generated
+         * source package root if they cannot be found in the source directory.
+         */
+        val CONFIG_FILES = arrayOf("package.json", "tsconfig.json", "babel.config.js")
+
+        /**
+         * Files to be copied from the reactor-ts submodule into the generated
+         * source directory.
+         */
+        val RUNTIME_FILES = arrayOf("cli.ts", "command-line-args.d.ts",
+            "command-line-usage.d.ts", "component.ts", "federation.ts", "reaction.ts",
+            "reactor.ts", "microtime.d.ts", "nanotimer.d.ts", "time.ts", "ulog.d.ts",
+            "util.ts")
+    }
+
+    init {
+        // Set defaults for federate compilation.
+        targetConfig.compiler = "gcc"
+        targetConfig.compilerFlags.add("-O2")
+    }
+
+    // Wrappers to expose GeneratorBase methods.
+    fun prw(builder: StringBuilder, text: Any) = pr(builder, text)
+    fun indentw(builder: StringBuilder) = indent(builder)
+    fun unindentw(builder: StringBuilder) = unindent(builder)
+
+    fun federationRTIPropertiesW() = federationRTIProperties
+
+    fun getTargetValueW(v: Value): String = getTargetValue(v)
+    fun getTargetTypeW(p: Parameter): String = getTargetType(p.inferredType)
+    fun getTargetTypeW(state: StateVar): String = getTargetType(state)
+    fun getTargetTypeW(a: Action): String = getTargetType(a)
+    fun getTargetTypeW(p: Port): String = getTargetType(p)
+    fun getTargetTypeW(t: Type): String = getTargetType(t)
+
+    fun getInitializerListW(state: StateVar): List<String> = getInitializerList(state)
+    fun getInitializerListW(param: Parameter): List<String> = getInitializerList(param)
+    fun getInitializerListW(param: Parameter, i: Instantiation): List<String> =
+        getInitializerList(param, i)
+    fun generateVarRefW(reference: VarRef): String =generateVarRef(reference)
+
+    /** Generate TypeScript code from the Lingua Franca model contained by the
+     *  specified resource. This is the main entry point for code
+     *  generation.
+     *  @param resource The resource containing the source code.
+     *  @param fsa The file system access (used to write the result).
+     *  @param context FIXME: Undocumented argument. No idea what this is.
+     */
+    override fun doGenerate(resource: Resource, fsa: IFileSystemAccess2,
```

Comment:
This is OK as a first port, but we should definitely break this down into smaller parts.

Reply:
Agreed! I added a TODO here to make sure that I follow up.